test_project/test_app/validators.py

```diff
@@ -0,0 +1,16 @@
+from django.core.exceptions import ValidationError
+
+
+def validate_ingredients(value):
```

Comment:
Is this really field-level validation, how it stands now?

Normally field-level validation simply checks the incoming `value` without much else. In this case, you instantiate a `Recipe` object, you process it, and check it. In essense, this is object-level validation. It's a subtle but important difference, and in my mind fails to communicate our intention here.

So, there are 2 options:
- We either don't do field-level validation at all, since it's actually out of the scope of this module. Object-level validation is.
- If we really want to show field-level validation, we do it differently. For example, check that the `value` is meaningful, without instantiating a `Recipe` object, but rather just using the `value` itself.

Reply:
I re-read the code and silly me I just did two validations twice 🤦‍♂ . So my mistakes here, I did field-validation through `validators` attribute, but then I stripped the value once again inside `clean()` method which is not necessary since the `ingredients` value is already validated on field-level validation.

Reply:
Done. And anyway, I choose options two. I still wanted to share how efficient `validators` attribute can be.

Reply:
Awesome work man!